File: ios/Classes/Convert.swift
import MapboxMaps
import Foundation

class Convert {
    class func interpretMapboxMapOptions(options: Any?, delegate: MapboxMapOptionsSink) {
        guard let options = options as? [String: Any] else { return }
        if let cameraTargetBounds = options["cameraTargetBounds"] as? [[[Double]]] {
            delegate.setCameraTargetBounds(bounds: CoordinateBounds.fromArray(cameraTargetBounds[0]))
        }
        if let compassEnabled = options["compassEnabled"] as? Bool {
            delegate.setCompassEnabled(enabled: compassEnabled)
        }
        if let scaleBarEnabled = options["scaleBarEnabled"] as? Bool {
            delegate.setScaleBarEnabled(enabled: scaleBarEnabled)
        }
        if let minMaxZoomPreference = options["minMaxZoomPreference"] as? [Double] {
            delegate.setMinMaxZoomPreference(min: minMaxZoomPreference[0], max: minMaxZoomPreference[1])
        }
        if let minMaxPitchPreference = options["minMaxPitchPreference"] as? [Double] {
            delegate.setMinMaxPitchPreference(min: minMaxPitchPreference[0], max: minMaxPitchPreference[1])
        }
        if let styleString = options["styleString"] as? String {
            delegate.setStyleString(styleString: styleString)
        }
        if let rotateGesturesEnabled = options["rotateGesturesEnabled"] as? Bool {
            delegate.setRotateGesturesEnabled(enabled: rotateGesturesEnabled)
        }
        if let scrollGesturesEnabled = options["scrollGesturesEnabled"] as? Bool {
            delegate.setScrollGesturesEnabled(enabled: scrollGesturesEnabled)
        }
        if let tiltGesturesEnabled = options["tiltGesturesEnabled"] as? Bool {
            delegate.setTiltGesturesEnabled(enabled: tiltGesturesEnabled)
        }
        if let trackCameraPosition = options["trackCameraPosition"] as? Bool {
            delegate.setTrackCameraPosition(trackCameraPosition: trackCameraPosition)
        }
        if let zoomGesturesEnabled = options["zoomGesturesEnabled"] as? Bool {
            delegate.setZoomGesturesEnabled(enabled: zoomGesturesEnabled)
        }
        if let myLocationEnabled = options["myLocationEnabled"] as? Bool {
            delegate.setMyLocationEnabled(enabled: myLocationEnabled)
        }
        if let myLocationTrackingMode = options["myLocationTrackingMode"] as? UInt, let trackingMode = MyLocationTrackingMode(rawValue: myLocationTrackingMode) {
            delegate.setMyLocationTrackingMode(myLocationTrackingMode: trackingMode)
        }
//        if let myLocationRenderMode = options["myLocationRenderMode"] as? Int, let renderMode = MyLocationRenderMode(rawValue: myLocationRenderMode) {
//            delegate.setMyLocationRenderMode(renderMode: renderMode)
//        }
        if let logoViewMargins = options["logoViewMargins"] as? [Double] {
            delegate.setLogoViewMargins(x: logoViewMargins[0], y: logoViewMargins[1])
        }
//        if let compassViewPosition = options["compassViewPosition"] as? UInt, let position = MGLOrnamentPosition(rawValue: compassViewPosition) {
//            delegate.setCompassViewPosition(position: position)
//        }
        if let compassViewMargins = options["compassViewMargins"] as? [Double] {
            delegate.setCompassViewMargins(x: compassViewMargins[0], y: compassViewMargins[1])
        }
        if let attributionButtonMargins = options["attributionButtonMargins"] as? [Double] {
            delegate.setAttributionButtonMargins(x: attributionButtonMargins[0], y: attributionButtonMargins[1])
        }
    }
    
    class func parseCameraUpdate(cameraUpdate: [Any], mapView: MapView) -> CameraOptions? {
            guard let type = cameraUpdate[0] as? String else { return nil }
            switch (type) {

            case "newCameraPosition":
                guard let cameraPosition = cameraUpdate[1] as? [String: Any] else { return nil }
                if let camera = mapView.camera,
                    let center = cameraPosition["target"] as? [Double],
                    let zoom = cameraPosition["zoom"] as? Double,
                    let bearing = cameraPosition["bearing"] as? Double,
                    let pitch = cameraPosition["pitch"] as? Double {
                    let cameraState = mapView.cameraState
                    return CameraOptions(
                        center: CLLocationCoordinate2D.fromArray(center),
                        padding: nil,
                        anchor: nil,
                        zoom: zoom,
                        bearing: bearing,
                        pitch: pitch
                    )
                }
            case "newLatLng":
                guard let coordinate = cameraUpdate[1] as? [Double] else { return nil }
                if let camera = mapView.camera {
                    return  CameraOptions(
                        center: CLLocationCoordinate2D.fromArray(coordinate),
                        padding: nil,
                        anchor: nil,
                        zoom: nil,
                        bearing: nil,
                        pitch: nil
                    )
                }
                /*
            case "newLatLngBounds":
                guard let bounds = cameraUpdate[1] as? [[Double]] else { return nil }
                guard let paddingLeft = cameraUpdate[2] as? CGFloat else { return nil }
                guard let paddingTop = cameraUpdate[3] as? CGFloat else { return nil }
                guard let paddingRight = cameraUpdate[4] as? CGFloat else { return nil }
                guard let paddingBottom = cameraUpdate[5] as? CGFloat else { return nil }
                return mapView.cameraThatFitsCoordinateBounds(MGLCoordinateBounds.fromArray(bounds), edgePadding: UIEdgeInsets.init(top: paddingTop, left: paddingLeft, bottom: paddingBottom, right: paddingRight))
            case "newLatLngZoom":
                guard let coordinate = cameraUpdate[1] as? [Double] else { return nil }
                guard let zoom = cameraUpdate[2] as? Double else { return nil }
                let camera = mapView.camera
                camera.centerCoordinate = CLLocationCoordinate2D.fromArray(coordinate)
                let altitude = getAltitude(zoom: zoom, mapView: mapView)
                return MGLMapCamera(lookingAtCenter: camera.centerCoordinate, altitude: altitude, pitch: camera.pitch, heading: camera.heading)
            case "scrollBy":
                guard let x = cameraUpdate[1] as? CGFloat else { return nil }
                guard let y = cameraUpdate[2] as? CGFloat else { return nil }
                let camera = mapView.camera
                let mapPoint = mapView.convert(camera.centerCoordinate, toPointTo: mapView)
                let movedPoint = CGPoint(x: mapPoint.x + x, y: mapPoint.y + y)
                camera.centerCoordinate = mapView.convert(movedPoint, toCoordinateFrom: mapView)
                return camera
            case "zoomBy":
                guard let zoomBy = cameraUpdate[1] as? Double else { return nil }
                let camera = mapView.camera
                let zoom = getZoom(mapView: mapView)
                let altitude = getAltitude(zoom: zoom+zoomBy, mapView: mapView)
                camera.altitude = altitude
                if (cameraUpdate.count == 2) {
                    return camera
                } else {
                    guard let point = cameraUpdate[2] as? [CGFloat], point.count == 2 else { return nil }
                    let movedPoint = CGPoint(x: point[0], y: point[1])
                    camera.centerCoordinate = mapView.convert(movedPoint, toCoordinateFrom: mapView)
                    return camera
                }
                 */
            case "zoomIn":
                if let camera = mapView.camera {
                    let cameraState = mapView.cameraState
                    return CameraOptions(center: nil, padding: nil, anchor: nil, zoom: cameraState.zoom + 1 , bearing: nil, pitch: nil)
                }
            case "zoomOut":
                if let camera = mapView.camera{
                    let cameraState = mapView.cameraState
                    return  CameraOptions(center: nil, padding: nil, anchor: nil, zoom: cameraState.zoom - 1, bearing: nil, pitch: nil)
                }
            case "zoomTo":
                guard let zoom = cameraUpdate[1] as? Double else { return nil }
                if let camera = mapView.camera {
                    return  CameraOptions(center: nil, padding: nil, anchor: nil, zoom: zoom, bearing: nil, pitch: nil)
                }
               
            case "bearingTo":
                guard let bearing = cameraUpdate[1] as? Double else { return nil }
                if let camera = mapView.camera {
                    return  CameraOptions(center: nil, padding: nil, anchor: nil, zoom: nil, bearing: bearing, pitch: nil)
                }
            case "pitchTo":
                guard let pitch = cameraUpdate[1] as? CGFloat else { return nil }
                if let camera = mapView.camera {
                    return CameraOptions(center: nil, padding: nil, anchor: nil, zoom: nil, bearing: nil, pitch: pitch)
                }

            default:
                print("\(type) not implemented!")
            }
            return nil
        }
    
    
    class func toPolygons(geometry: [[[Double]]]) -> [Polygon] {
        var polygons:[Polygon] = []
        for lineString in geometry {
            var linearRing: [CLLocationCoordinate2D] = []
            for coordinate in lineString {
                linearRing.append(CLLocationCoordinate2DMake(coordinate[0], coordinate[1]))
            }
            let polygon = Polygon([linearRing])
            polygons.append(polygon)
        }
        return polygons
    }
    
    class func serializeJSON(from object:Any) -> String? {
        do {
            let isArray = object is NSArray
            let isDictionary = object is NSDictionary
            if ( !isArray && !isDictionary ) {
                return nil
            }

            guard let data = try? JSONSerialization.data(withJSONObject: object, options: []) else {
                return nil
            }
            return String(data: data, encoding: String.Encoding.utf8)
        } catch {
            return nil
        }
    }
    
    class func deserializeJSON(text: String) -> [String: Any]? {
        if let data = text.data(using: .utf8) {
            do {
                return try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any]
            } catch {
                print(error.localizedDescription)
            }
        }
        return nil
    }
    
    class func getVectorURLTemplated( properties: [String: Any]) -> [String] {
        var urls: [String] = []
        for (propertyName, propertyValue) in properties {
            switch propertyName {
                case "tiles":
                    let uriList = propertyValue as? [String]
                    if uriList != nil {
                        for uri in uriList! {
                            urls.append(uri.removingPercentEncoding!);
                        }
                    }
                    
                default:
                    break
            }
        }
        return urls
    }
}

File: android/src/main/kotlin/com/mapbox/mapboxgl/Convert.kt
// Copyright 2018 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.
package com.mapbox.mapboxgl

import com.mapbox.geojson.Point
import com.mapbox.geojson.Polygon
import com.mapbox.maps.CameraOptions
import com.mapbox.maps.CameraState
import com.mapbox.maps.MapboxMap
import java.util.*

/**
 * Conversions between JSON-like values and MapboxMaps data types.
 */
internal object Convert {
    private const val TAG = "Convert"

    private fun toBoolean(o: Any): Boolean {
        return o as Boolean
    }

    fun toCameraPosition(o: Any?): CameraOptions {
        val data = toMap(o)
        val builder = CameraOptions.Builder()
        data?.let { data ->
            builder.bearing(data["bearing"] as Double?)
            (data["target"] as List<Double>?)?.let { center ->
                builder.center(Point.fromLngLat(center[1], center[0]))
            }

            builder.pitch(data["tilt"] as Double?)
            builder.zoom(data["zoom"] as Double?)
        }
        return builder.build()
    }

    fun toAnnotationOrder(o: Any?): List<String?> {
        val data = toList(o)
        val annotations: MutableList<String?> = arrayListOf();
        for (index in data!!.indices) {
            annotations.add(toString(data!![index]!!))
        }
        return annotations
    }

    fun toCameraUpdate(o: Any?, mapboxMap: MapboxMap): CameraOptions? {
        val data = toList(o)
        var action = toString(data!![0]!!);

        var cameraOptions = CameraOptions.Builder();

        return when (action) {
            "newCameraPosition" -> toCameraPosition(data[1])
//            "newLatLng" -> CameraUpdateFactory.newLatLng(toLatLng(data[1]))
//            "newLatLngBounds" -> CameraUpdateFactory.newLatLngBounds(toLatLngBounds(data[1])!!, toPixels(data[2]!!, density),
//                    toPixels(data[3]!!, density), toPixels(data[4]!!, density), toPixels(data[5]!!, density))
//            "newLatLngZoom" -> CameraUpdateFactory.newLatLngZoom(toLatLng(data[1]), toFloat(data[2]).toDouble())
//            "scrollBy" -> {
//                mapboxMap!!.scrollBy(
//                        toFractionalPixels(data[1]!!, density),
//                        toFractionalPixels(data[2]!!, density)
//                )
//                null
//            }
//            "zoomBy" -> if (data.size == 2) {
//                CameraUpdateFactory.zoomBy(toFloat(data[1]).toDouble())
//            } else {
//                CameraUpdateFactory.zoomBy(toFloat(data[1]).toDouble(), toPoint(data[2]!!, density))
//            }
            "zoomIn" -> cameraOptions.zoom(mapboxMap.cameraState.zoom + 1).build()
            "zoomOut" -> cameraOptions.zoom(mapboxMap.cameraState.zoom - 1).build()
            "zoomTo" -> cameraOptions.zoom(data[1] as Double).build()
            "bearingTo" -> cameraOptions.bearing(data[1] as Double).build()
            "pitchTo" -> cameraOptions.pitch(data[1] as Double).build()
            else -> throw IllegalArgumentException("Cannot interpret $o as CameraUpdate")
        }
    }

    private fun toDouble(o: Any): Double {
        return (o as Number).toDouble()
    }

    private fun toFloat(o: Any?): Float {
        return (o as Number?)?.toFloat() ?: (0.0).toFloat()
    }

    private fun toInt(o: Any): Int {
        return (o as Number).toInt()
    }

    fun toJson(position: CameraState?): Any? {
        if (position == null) {
            return null
        }
        val data: MutableMap<String, Any> = HashMap()
        data["bearing"] = position.bearing
        data["target"] = toJson(position.center)
        data["pitch"] = position.pitch
        data["zoom"] = position.zoom
        return data
    }

    private fun toJson(latLng: Point): Any {
        return listOf(latLng.latitude(), latLng.longitude())
    }

//    private fun toLatLng(o: Any?): LatLng {
//        val data = toList(o)
//        return LatLng(toDouble(data!![0]!!), toDouble(data[1]!!))
//    }
//
//    private fun toLatLngBounds(o: Any?): LatLngBounds? {
//        if (o == null) {
//            return null
//        }
//        val data = toList(o)
//        val boundsArray = arrayOf(toLatLng(data!![0]), toLatLng(data[1]))
//        val bounds = listOf(*boundsArray)
//        val builder = LatLngBounds.Builder()
//        for(bound in bounds) {
//            builder.include(bound)
//        }
//        return builder.build()
//    }

    private fun toList(o: Any?): List<*>? {
        return o as List<*>?
    }

    fun toMap(o: Any?): Map<*, *>? {
        return o as Map<*, *>?
    }

    private fun toFractionalPixels(o: Any, density: Float): Float {
        return toFloat(o) * density
    }

    fun toPixels(o: Any, density: Float): Int {
        return toFractionalPixels(o, density).toInt()
    }

    private fun toPoint(o: Any, density: Float): android.graphics.Point {
        val data = toList(o)
        return android.graphics.Point(toPixels(data!![0]!!, density), toPixels(data[1]!!, density))
    }

    private fun toString(o: Any): String {
        return o as String
    }

    fun interpretMapboxMapOptions(o: Map<String, Any?>, sink: MapboxMapOptionsSink) {
        toMap(o)?.let { data ->
//            val cameraTargetBounds = data["cameraTargetBounds"]
//            if (cameraTargetBounds != null) {
//                val targetData = toList(cameraTargetBounds)
//                sink.setCameraTargetBounds(toLatLngBounds(targetData!![0]))
//            }
            val compassEnabled = data["compassEnabled"]
            if (compassEnabled != null) {
                sink.setCompassEnabled(toBoolean(compassEnabled))
            }
            val scaleBarEnabled = data["scaleBarEnabled"]
            if (scaleBarEnabled != null) {
                sink.setScaleBarEnabled(toBoolean(scaleBarEnabled))
            }
            val styleString = data["styleString"]
            if (styleString != null) {
                sink.setStyleString(toString(styleString))
            }
            val minMaxZoomPreference = data["minMaxZoomPreference"]
            if (minMaxZoomPreference != null) {
                val zoomPreferenceData = toList(minMaxZoomPreference)!!
                sink.setMinMaxZoomPreference( //
                    zoomPreferenceData[0] as Double?,
                    zoomPreferenceData[1] as Double?
                )
            }
            val minMaxPitchPreference = data["minMaxPitchPreference"]
            if (minMaxPitchPreference != null) {
                val pitchPreferenceData = toList(minMaxPitchPreference)!!
                sink.setMinMaxPitchPreference( //
                    pitchPreferenceData[0] as Double?,
                    pitchPreferenceData[1] as Double?
                )
            }
            val rotateGesturesEnabled = data["rotateGesturesEnabled"]
            if (rotateGesturesEnabled != null) {
                sink.setRotateGesturesEnabled(toBoolean(rotateGesturesEnabled))
            }
            val scrollGesturesEnabled = data["scrollGesturesEnabled"]
            if (scrollGesturesEnabled != null) {
                sink.setScrollGesturesEnabled(toBoolean(scrollGesturesEnabled))
            }
            val tiltGesturesEnabled = data["tiltGesturesEnabled"]
            if (tiltGesturesEnabled != null) {
                sink.setPitchGesturesEnabled(toBoolean(tiltGesturesEnabled))
            }
            val trackCameraPosition = data["trackCameraPosition"]
            if (trackCameraPosition != null) {
                sink.setTrackCameraPosition(toBoolean(trackCameraPosition))
            }
            val zoomGesturesEnabled = data["zoomGesturesEnabled"]
            if (zoomGesturesEnabled != null) {
                sink.setZoomGesturesEnabled(toBoolean(zoomGesturesEnabled))
            }
            val myLocationEnabled = data["myLocationEnabled"]
            if (myLocationEnabled != null) {
                sink.setMyLocationEnabled(toBoolean(myLocationEnabled))
            }
            val myLocationTrackingMode = data["myLocationTrackingMode"]
            if (myLocationTrackingMode != null) {
                sink.setMyLocationTrackingMode(toInt(myLocationTrackingMode))
            }
            val myLocationRenderMode = data["myLocationRenderMode"]
            if (myLocationRenderMode != null) {
                sink.setMyLocationRenderMode(toInt(myLocationRenderMode))
            }
            val logoPosition = data["logoViewPosition"]
            if (logoPosition != null) {
                sink.setLogoPosition(toInt(logoPosition))
            }
            val logoViewMargins = data["logoViewMargins"]
            if (logoViewMargins != null) {
                val logoViewMarginsData = toList(logoViewMargins)
                sink.setLogoViewMargins(toInt(logoViewMarginsData!![0]!!), toInt(logoViewMarginsData[1]!!))
            }
            val compassPosition = data["compassViewPosition"]
            if (compassPosition != null) {
                sink.setCompassViewPosition(toInt(compassPosition))
            }
            val compassViewMargins = data["compassViewMargins"]
            if (compassViewMargins != null) {
                val compassViewMarginsData = toList(compassViewMargins)
                sink.setCompassViewMargins(toInt(compassViewMarginsData!![0]!!), toInt(compassViewMarginsData[1]!!))
            }
            val attributionButtonPosition = data["attributionButtonPosition"]
            if (attributionButtonPosition != null) {
                sink.setAttributionButtonPosition(toInt(attributionButtonPosition))
            }
            val attributionButtonMargins = data["attributionButtonMargins"]
            if (attributionButtonMargins != null) {
                val attributionButtonMarginsData = toList(attributionButtonMargins)
                sink.setAttributionButtonMargins(toInt(attributionButtonMarginsData!![0]!!), toInt(attributionButtonMarginsData[1]!!))
            }
        }
    }
}